### DISCUSSION TITLE
Possible to input the required kwh

### DISCUSSION THREAD

**Discussion by efauske** [elden1337/hass-peaq#181] [Ideas]
Hi!

I’ve just started using PeaqEV and wonder if it is possible to input the required number of kWh you want to charge in a session?

Either define a sensor reporting car battery % and total capacity, or just an numeric value for the # of kWh I want to transfer to the car.

This could help peaqev to make a better decision on when to administer the charge?

Anyways, thanks for a great addon!!

**Comment by elden1337** [comment]
Hello there!
Actually there is a function witin peaqev called scheduler which aims to let you do that. However, ive started on a scheduler 2.0 because the current one is buggy.

**Comment by elden1337** [comment]
But, there will be an issue keeping three conditions at bay simultaneously; peak, price and charge. 

Currently i ignore charge and let peaqev charge in whatever it can given price and peak. 

So in this situation its probable that either the peak will be slightly raised, or charging at "not cheapest" hours necessary. Or both. 

Lets see what the calculations come up with

**Comment by efauske** [comment]
Ah, now I found the topic in the Wiki! Thanks, it was just what I was looking for :) (sorry for not looking more closely before asking...)

So the current scheduler ignores the amount of kwh in the "charge_amount"? I appreciate that it's hard (impossible) to fulfill all three conditions. All the while you don't have any control of when other devices consume power, you'll never be able to fulfill all of them.

**Reply by elden1337** [reply]
The idea behind the scheduler currently is to give a range with the desired amount. It's not supposed to breach peak (if not explicitly stated in the call by the user), and then lay out a pattern of the cheapest hours during that "night" (which it usually is). 
However, i have realized it recalculates constantly and therefore isn't very lean towards the desires. 

The new one is supposed to be a set where you can tell peaq that you wish to charge 10kwh every weeknight, or 30kwh every sunday etc. 
Also, since i'm logging the prices and 24h-consumption (locally not for me just for you) i aim to try and create an auto-scheduler based on the patterns in your previous sessions.

But, time is scarce so we'll see when we get there ;).

**Reply by elden1337** [reply]
Ie, I would be catious about using the current scheduler since i know it is buggy.

**Reply by efauske** [reply]
Thanks for the heads up on current scheduler!

Would it be possible for peaq to take the current StateOfCharge of the car in %, along with total capacity (i.e 40kwh) and then automatically find out how many kwh's are required for the current session? Also perhaps with a time specifying when you'd like the charge to be completed by. 

If peaq keeps statistics of house consumption (excluding car), it will after a while be able to know more about how much is spent by other devices at different times of day (or especially for car charging - night), and it could use that to make an educated guess on how much it would be able to charge given expected peak levels. 

The input can include a "minimum StateOfCharge % target" as well, because it may not be that it is required to fill to 100% at every charge.

I see that adding more variables makes the algorithm more complex, but I'm just brainstorming :)

**Reply by elden1337** [reply]
Yes, and I do have tickets for the integrations ive found (in issues) but for me personally it's not a prioritized set. 
However, I would not object if someone would contribute and do it and merge to the repo ;)

**Reply by efauske** [reply]
I totally get that :) 

The current state of Peaq works well for me. I only have a 40kwh Leaf, so it is usually no problem with it not getting enough charge, event running at 50% of the charger capacity.

Thanks again for spending the time creating it :)

**Comment by elden1337** [comment]
This is taken care of in 2.10.0
